src/Features/Core/Portable/Completion/Providers/Snippets/AbstractSnippetCompletionProvider.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers.Snippets
+{
+    internal abstract class AbstractSnippetCompletionProvider : CommonCompletionProvider
+    {
+        private readonly SyntaxAnnotation _cursorAnnotation = new();
+        private readonly SyntaxAnnotation _reformatSnippetAnnotation = new();
+
+        protected abstract int GetTargetCaretPosition(SyntaxNode caretTarget);
+        protected abstract Task<Document> GenerateDocumentWithSnippetAsync(Document document, CompletionItem completionItem, CancellationToken cancellationToken);
+        protected abstract Task<SyntaxNode> GetAnnotatedSnippetRootAsync(Document document, CompletionItem completionItem,
+            SyntaxAnnotation reformatAnnotation, CancellationToken cancellationToken);
+        protected abstract Task<SyntaxNode> GetAnnotationForCursorAsync(Document document, CompletionItem completionItem,
+            SyntaxAnnotation cursorAnnotation, CancellationToken cancellationToken);
```

Comment:
not sure how to read this method.  it says GetAnnotation but returns a syntaxnode.